blogposts/2021/optimizing-a-code-intelligence-commit-graph.md

```diff
@@ -0,0 +1,699 @@
+
+```sql
+(l.direction = 'A' AND c.commit = l.parent_commit) OR (l.direction = 'D' AND c.parent_commit = l.commit)
+```
+
+To reduce the cost of this index scan, we rewrote the recursive term within the table expression above (the second term of the `UNION`). Instead of using a `SELECT` with an `OR` condition (which had trouble using our existing set of indexes), we use `UNION` to combine the results matching each side of the `OR` clause.
+
+```sql
+SELECT c.*, l.distance + 1, l.direction
+FROM lineage l
+JOIN commits_with_lsif_data_markers c
+ON c.repository = l.repository AND l.direction = 'A' AND c."commit" = l.parent_commit
+
+UNION
+
+SELECT c.*, l.distance + 1, l.direction
+FROM lineage l
+JOIN commits_with_lsif_data_markers c
+ON c.repository = l.repository AND l.direction = 'D' AND c.parent_commit = l."commit"
+```
+
+The query plan for this new query, shown below, seems more complex at first glance. However, this query is **drastically** more efficient. The same input that required 330ms to evaluate now takes under 1ms to evaluate.
+
+![super fast query plan](https://sourcegraphstatic.com/blog/commit-graph-optimizations/super-fast.png)
+
+Even though we're now executing a greater number of steps, each step can be evaluated efficiently. The old inefficient index scan has been broken into two different index scans: one that traverses the ancestor direction, and another that traverses the descendant direction. The simple conditionals in each query can be evaluated _without an index filter_ by each of the multicolumn indexes above. Instead of pulling back the entire commit graph on each iteration of the table expression, we pull exactly the set of rows that need to be added to the working table.
+
+This change makes the query efficient enough that we no longer have to worry about caching results, as running it on every request only contributes a negligible amount of time. Thus solving the problem [once and for all](https://www.youtube.com/watch?v=IjmtVKOAHPM)!
+
+## Scalability improvements
+
+Well, it solves the problem once and for all _at a particular scale_. There is an entire class of enterprise customers we would like to target to benefit from this feature as well, except the speed at which the code moves is a huge obstacle to overcome using this technique of calculating visible uploads on demand.
+
+In order to prevent queries which must travel a large distance over the commit graph before finding a visible upload from timing out or affecting the stability of the database, we still keep a hard limit on the number of commits we will visit during the graph traversal. The benefit of this limit is we bring stability to the Sourcegraph instance by limiting the maximum load a single query can put on the database. The downside of this limit is that there are many shapes of commit graphs and data markers that will fail to find a visible upload, even if the distance is not too large.
```

Comment:
```suggestion
In order to prevent queries which must travel a large distance over the commit graph before finding a visible upload from timing out or affecting the stability of the database, we still keep a hard limit on the number of commits we will visit during the graph traversal. The benefit of this limit is stability to the Sourcegraph instance by limiting the maximum load a single query can put on the database. The downside of this limit is that there are many shapes of commit graphs and data markers that will fail to find a visible upload, even if the distance is not too large.
```